### PR TITLE
Coalesce graph editor view invalidations

### LINKED ISSUE #6630: Performance regression on todays Nightly build in comparison to 2023.5.9 Nightly build
### Discord username

_No response_

### What type of issue is this?

Permanent – Occurring repeatably

### Is this issue blocking you from using Enso?

- [X] Yes, I can't use Enso because of this issue.

### Is this a regression?

- [ ] Yes, previous version of Enso did not have this issue.

### What issue are you facing?

over 2x performance drop on starting the project

Nightly 2023.5.10:

https://github.com/enso-org/enso/assets/108271897/d39c5c4d-a3ec-4605-bb1d-0b980eff5897

Nightly 2023.5.9:

https://github.com/enso-org/enso/assets/108271897/2c98f3b7-eef7-4c38-b044-8804476d3ada

### Expected behaviour

Enso performing at least at the level from the previous build.

### How we can reproduce it?

[Unnamed_5.zip](https://github.com/enso-org/enso/files/11441349/Unnamed_5.zip)

launch on today and yesterday Nightly builds

### Screenshots or screencasts

_No response_

### Enso Version

2023.5.10 nightly

### Browser or standalone distribution

Standalone distribution

### Browser Version or standalone distribution

standalone

### Operating System

Windows

### Operating System Version

Win11pro 22H2 22621.1555

### Hardware you are using

12th Gen Intel(R) Core(TM) i9-12900HK / RTX3060 Laptop / Nvidia Drivers 531.68

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

Only invalidate the graph editor view at most once per frame. On develop, this saves about 70ms (2%). Testing a recent backend without #6755 as a stress-test, this saves about 5s (45%). This reflects better scalability to large numbers of `SuggestionUpdate` messages.

Fixes #6630.

### Important Notes

- Also fix intermittent profiling failures occurring since the introduction of microtasks.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] ~The documentation has been updated, if necessary.~
- [ ] ~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [ ] ~Unit tests have been written where possible.~
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: lib/rust/ensogl/core/src/animation/loops.rs [hunk 1]
```diff
@@ -324,10 +324,7 @@ fn on_frame_closure(
             .then(move || on_after_animations.emit(time_info))
             .then(move || frame_end.emit(time_info))
             .then(move || on_before_layout.emit(time_info))
-            .then(move || {
-                on_before_rendering.emit(time_info);
-                drop(_profiler);
-            })
```
Comment (farmaazon) [inline lib/rust/ensogl/core/src/animation/loops.rs]: Is it intentional? Just making sure, as I don't know this part, and this change wasn't in PR's description.

**Review by farmaazon — APPROVED** [review]

**Review by wdanilo — APPROVED** [review]

**Comment by wdanilo** [comment]
A question – why does it fix https://github.com/enso-org/enso/issues/6630? Was that regression introduced by invalidating the graph editor multiple times per frame and this PR fixes that, or something else happened?

File: lib/rust/ensogl/core/src/animation/loops.rs (hunk 1 above)
Comment (kazcw) [reply lib/rust/ensogl/core/src/animation/loops.rs]: Thanks for noticing--fixed the description.

**Comment by kazcw** [comment]
> A question – why does it fix #6630? Was that regression introduced by invalidating the graph editor multiple times per frame and this PR fixes that, or something else happened?

I'll have a look at the increased cost of view initialization when #6630 started and see if there's much room for improvement there.

**Comment by kazcw** [comment]
We're spending 14% (475ms) of `Orders` startup time in 64 calls to widgets' `rebuild_tree`.

image

While I'm sure we could do better than 7ms per tree, this project only has 12 nodes. We shouldn't need to build 64 widget trees to open it. #6802 will probably help with this: When graph execution changes the data attached to `SpanTree` nodes instead of changing the whole `SpanTree` structure, it will be easier to update only widgets that have changed.

**Comment by kazcw** [comment]
Actually, widget rebuild is already incremental via a `StableSpanIdentity` mechanism. Whether #6802 would help with performance here would take more research, though it would help with simplicity.

**Comment by Frizi** [comment]
@kazcw About the widget updates, I don't think that the amount of updates here is in any way alarming. And as you noticed, most of those updates should indeed be done incrementally, due to stable identities. The problem is that creating UI components in general is very expensive right now, partly due to the overhead of creating the display object hierarchy and each component's FRP network (to be more specific, due to the huge amount of tiny allocations needed to do that). I expect that this will improve drastically with the upcoming FRP changes.

Separating the SpanTree (or expression data in general) from execution context data won't necessarily improve the situation, because the work to update the widgets fundamentally has to happen anyway. We would just have two separate streams of data that the tree has to react to. Also, due to how the node expression update is handled, updates with unchanging span tree actually shouldn't even cause the widget tree to be rebuilt. Those unchanged updates are rejected at the graph presenter level.

https://github.com/enso-org/enso/blob/fe0a06dcb50a82f752382882b6592911f6e3b65d/app/gui/src/presenter/graph/state.rs#L464

Interestingly, the way this is checked is another area for improvement. Building the span-trees only to deep compare them and figure out nothing changed is incredibly wasteful.

**Comment by farmaazon** [comment]
## QA Report: :red_circle: 

Detaching connections going to self port from their source and connecting to another source triggers weird behavior:

https://github.com/enso-org/enso/assets/3919101/06e86d9d-bdd0-43c6-a27f-c1117b59dbf0

Often I cannot connect to this self port afterward.

Now, I have actually seen the same issue on the develop once, but in this PR it became 100% reproducible.

**Comment by sylwiabr** [comment]
@farmaazon can you re-upload the video?

**Comment by farmaazon** [comment]
Done. Sorry, I missed it hadn't uploaded.

**Comment by kazcw** [comment]
That is #6772, which I don't think is intermittent, except for some timing-sensitivity. It depends on: The target node and port (typically detaching a port that is the receiver of a method, as in your example, causes it), and how long the edge is detached (it breaks when the graph re-evaluates). In a graph similar to your example, I can reproduce it 100% of the time on `develop` @ b353b35.

**Comment by farmaazon** [comment]
You are right, it's far more reducible on develop than I've seen during my QA (perhaps I was just lucky). So **the QA turns :green_circle: **

**Pull request merged**